### ISSUE TITLE
Infer extension-requirements for nested DFGs

### ISSUE THREAD

**Issue by acl-cqc** [CQCL/hugr#640]
Nodes such as DFG, CFG, BasicBlock etc. are created wth a FunctionType that specifies a delta: https://github.com/CQCL/hugr/blob/1a07cd9aca60aa75257bdb297d7442d9529765fb/src/types/signature.rs#L25

This must be specified explicitly at creation time - one cannot say "None" as one can for `input_extensions`. However it would be nicer if one could, such that the inference algorithm then infers it, which I believe is possible as follows.

(This would require some change to the FunctionType/NodeType/DFG structures to allow specifying an Option, and then mutating it when inference is complete, in discussion below)

During inference, we both 1. constraint the input+output variables to match those at the Input and Output nodes of the DFG: https://github.com/CQCL/hugr/blob/1a07cd9aca60aa75257bdb297d7442d9529765fb/src/extension/infer.rs#L283-L288 and 2. constrain the outputs to be the input plus the specified delta: https://github.com/CQCL/hugr/blob/1a07cd9aca60aa75257bdb297d7442d9529765fb/src/extension/infer.rs#L314-L318

If we simplify avoid the second constraints in such cases, the constraints with the Input/Output nodes should be sufficient to calculate both. We can then compute the delta as the difference between the two. (If this last seems dubious - indeed the delta could also include any subset of the *inputs* - then this is no worse than at present, where such a delta specified by the programmer would be accepted by inference and validation. We could instead move to a system where the Input node also has *empty* input extensions, but that's another issue.)

**Comment by acl-cqc** [comment]
Thus, two possible solutions:
* Recursively infer delta from the nested subgraph first, use that in the outer graph. This won't handle functions containing recursive calls (you need the delta from the function body to infer around the call node) - user-provided annotations on functions?
* Try to generalize "Plus" constraint to merge two variables. This might magically work or break everything, not sure.

**Comment by acl-cqc** [comment]
> This would require some change to the FunctionType/NodeType/DFG structures to allow specifying an Option, and then mutating it when inference is complete

So it's worth thinking about this because it could be quite disruptive. There are several possible methods, including:
1. Change `FunctionType` to store an `Option<ExtensionSet>`. Then validation checks all these (including those nested *within* port/edge types) are non-None. *However* it's not clear that `None` is useful in many places, e.g. outside of a node's `signature()`. Moreover having to deal with `Option` rather than `ExtensionSet` *everywhere* one has a FunctionType makes this seem like the most disruptive method.
2. Separate Node `signature()` from FunctionType; the former has an Option, the latter a non-optional ExtensionSet. Given we are working on #1094 (e.g. #1138) this seems possible but could be very complex.
3. Leave all signatures/FunctionTypes containing an ExtensionSet, but allow inference to mutate it (!). There are are few further variants...
    * Allow any container node's `extension_reqs` to be mutated. This means that extension-delta-inference will never fail, but seems a bit wild.
    * Allow, say, only container `extension_reqs` that are currently empty to be inferred.
    * Restrict inference to e.g. only increasing `extension_reqs`, or only reducing them.
    * As a variant on the latter, suppose some kind of special ExtensionId wildcard/universal-set, which could be seen as meaning either "all possible extensions" or "please infer me". An ExtensionSet containing both the special please-infer *and* other ExtensionIds coould mean "at least the others, possibly more". We might want to make the *default* ExtensionSet contain only the wildcard, and (even more likely) make the builder default to using just the wildcard.
4. Remove `extension_reqs` from FunctionType, so the latter is just inputs/outputs; then a `Type(Enum)::Function` stores both a FunctionType (in/out) *and* a non-optional ExtensionSet; whereas a node has a "signature() -> FunctionType" and a separate "extension_delta: Optional<ExtensionSet>" that can be set (from None) by inference. [Yes, this is a different spin on 2., but it might be a good way of decoupling the delta from the FunctionType-does/doesn't-have-rowvariables question.]

**Comment by acl-cqc** [comment]
Closing after #1195, front-end/builder changes can be separate issue (and #1237 is a separate issue!)
